extension/flat_tensor/serialize/serialize.cpp

```diff
@@ -0,0 +1,138 @@
+    // Returns the padding required to align `offset` to `alignment`.
+    size_t remainder = offset % alignment;
+    if (remainder != 0) {
+        return alignment - remainder;
+    }
+    return 0;
+  }
+
+  size_t aligned_size(size_t input_size, size_t alignment) {
+    // Returns input_size padded up to the next whole multiple of alignment.
+    return input_size + padding_required(input_size, alignment);
+  }
+
+  void serialize_nulls(std::ostream& out, size_t num_bytes) {
+    for (size_t i = 0; i < num_bytes; i++) {
+      out.write("\0", 1);
+    }
+  }
+}
+
+ET_EXPERIMENTAL runtime::Error save_ptd(
+  const std::string& path,
+  const std::map<std::string, exec_aten::Tensor>& tensor_map,
+  const size_t tensor_alignment
+) {
+  // Assert the system is little endian
+  int n = 1;
+  if(*(char *)&n != 1) {
+    return runtime::Error::NotSupported;
+  }
+  // Create File
+  std::ofstream file;
+  file.open(path);
+  runtime::Error e = save_ptd(file, tensor_map, tensor_alignment);
+  file.close();
+  return e;
+}
+
+ET_EXPERIMENTAL runtime::Error save_ptd(
+  std::ostream& out,
+  const std::map<std::string, exec_aten::Tensor>& tensor_map,
+  const size_t tensor_alignment
+) {
+  // Assert the system is little endian
+  int n = 1;
+  if(*(char *)&n != 1) {
+    return runtime::Error::NotSupported;
+  }
+  // Create flatbuffer
+  flatbuffers::FlatBufferBuilder builder;
+
+  std::vector<flatbuffers::Offset<::flat_tensor::TensorMetadata>> tensors;
+  std::vector<flatbuffers::Offset<::flat_tensor::DataSegment>> buffers;
+
+  // Serialize the tensors.
+  size_t total_segment_size = 0;
+  for (const auto& [name, tensor] : tensor_map) {
+    auto name_offset = builder.CreateString(name);
+    // Serialize the tensor metadata.
+    auto tensor_metadata = ::flat_tensor::CreateTensorMetadata(
+      builder,
+      name_offset,
+      static_cast<executorch_flatbuffer::ScalarType>(tensor.scalar_type()),
+      builder.CreateVector(tensor.sizes().data(), tensor.sizes().size()),
+      builder.CreateVector(tensor.dim_order().data(), tensor.dim_order().size()),
+      0, // segment index
+      total_segment_size);
+    tensors.push_back(tensor_metadata);      // Precalculate the size of the data blob
+    total_segment_size += aligned_size(tensor.nbytes(), tensor_alignment);
+  }
+  // Only have one segment
+  buffers.push_back(::flat_tensor::CreateDataSegment(builder, 0, total_segment_size));
+  
+  auto flat_tensor = CreateFlatTensor(builder, internal::VERSION, tensor_alignment, builder.CreateVector(tensors), builder.CreateVector(buffers));
+  builder.Finish(flat_tensor);  // Our flatbuffer is created now.
+
+
+  // Calculate flatbuffer padding.
+  auto padded_flatbufer_size = aligned_size(builder.GetSize(), tensor_alignment);
+  auto padded_header_size = aligned_size(internal::HEADER_EXPECTED_LENGTH, tensor_alignment);
+
+  // Serialize header
+  out.write(internal::HEADER_MAGIC, sizeof(internal::HEADER_MAGIC));
```

Comment:
This is arranged differently from PTE files. I hid the extended header inside the flatbuffer data so that a PTE file can still be read by standard flatbuffer tools, which has been really useful.

See https://github.com/pytorch/executorch/blob/2411c9913309914ce8dd6db7e17b96ef85907bd6/exir/_serialize/_program.py#L55 for that logic